File: app/repositories/tenants.py
"""Repositórios de tenant/user/membership — todo o SQL dessas tabelas.

São a base do modelo multi-tenant (ponto 19) e do RBAC (ponto 8): o tenant é
dono dos agentes, o usuário se vincula ao tenant por uma membership com papel.
"""

from app.db import read_connection, transaction
from app.domain import Membership, Tenant, User


class TenantRepository:
    def get(self, tenant_id: str) -> Tenant | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM tenants WHERE id = ?", (tenant_id,)).fetchone()
            return Tenant.from_row(row) if row else None

    def get_by_api_key(self, api_key: str) -> Tenant | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM tenants WHERE api_key = ?", (api_key,)).fetchone()
            return Tenant.from_row(row) if row else None

    def list(self) -> list[Tenant]:
        with read_connection() as conn:
            rows = conn.execute("SELECT * FROM tenants ORDER BY created_at").fetchall()
            return [Tenant.from_row(r) for r in rows]

    def exists(self, tenant_id: str) -> bool:
        with read_connection() as conn:
            return conn.execute("SELECT 1 FROM tenants WHERE id = ?", (tenant_id,)).fetchone() is not None

    def insert(self, tenant_id: str, name: str, api_key: str) -> None:
        with transaction() as conn:
            conn.execute(
                "INSERT INTO tenants (id, name, api_key) VALUES (?, ?, ?)",
                (tenant_id, name, api_key),
            )

    def delete(self, tenant_id: str) -> bool:
        with transaction() as conn:
            return conn.execute("DELETE FROM tenants WHERE id = ?", (tenant_id,)).rowcount > 0


class UserRepository:
    def get(self, user_id: str) -> User | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM users WHERE id = ?", (user_id,)).fetchone()
            return User.from_row(row) if row else None

    def get_by_api_key(self, api_key: str) -> User | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM users WHERE api_key = ?", (api_key,)).fetchone()
            return User.from_row(row) if row else None

    def get_by_email(self, email: str) -> User | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM users WHERE email = ?", (email,)).fetchone()
            return User.from_row(row) if row else None

    def insert(self, user_id: str, email: str, name: str, api_key: str) -> None:
        with transaction() as conn:
            conn.execute(
                "INSERT INTO users (id, email, name, api_key) VALUES (?, ?, ?, ?)",
                (user_id, email, name, api_key),
            )


class MembershipRepository:
    def get(self, tenant_id: str, user_id: str) -> Membership | None:
        with read_connection() as conn:
            row = conn.execute(
                "SELECT * FROM memberships WHERE tenant_id = ? AND user_id = ?",
                (tenant_id, user_id),
            ).fetchone()
            return Membership.from_row(row) if row else None

    def list_for_tenant(self, tenant_id: str) -> list[Membership]:
        with read_connection() as conn:
            rows = conn.execute(
                "SELECT * FROM memberships WHERE tenant_id = ? ORDER BY role, user_id",
                (tenant_id,),
            ).fetchall()
            return [Membership.from_row(r) for r in rows]

    def upsert(self, tenant_id: str, user_id: str, role: str) -> None:
        with transaction() as conn:
            conn.execute(
                "INSERT INTO memberships (tenant_id, user_id, role) VALUES (?, ?, ?) "
                "ON CONFLICT(tenant_id, user_id) DO UPDATE SET role = excluded.role",
                (tenant_id, user_id, role),
            )

    def delete(self, tenant_id: str, user_id: str) -> bool:
        with transaction() as conn:
            return conn.execute(
                "DELETE FROM memberships WHERE tenant_id = ? AND user_id = ?",
                (tenant_id, user_id),
            ).rowcount > 0

    def count_owners(self, tenant_id: str) -> int:
        with read_connection() as conn:
            return conn.execute(
                "SELECT COUNT(*) FROM memberships WHERE tenant_id = ? AND role = 'owner'",
                (tenant_id,),
            ).fetchone()[0]

File: src/server/repositories/tenants.ts
/**
 * Repositórios de tenant/user/membership (porta `app/repositories/tenants.py`).
 * Base do modelo multi-tenant e do RBAC. Todo o SQL dessas tabelas vive aqui.
 */

import { and, asc, count, eq } from "drizzle-orm";

import { db } from "../db/client";
import { memberships, tenants, users } from "../db/schema";
import type { Membership, MemberRole, Tenant, User } from "../domain";
import { toIso } from "./util";

type TenantRow = typeof tenants.$inferSelect;
type UserRow = typeof users.$inferSelect;
type MembershipRow = typeof memberships.$inferSelect;

function toTenant(r: TenantRow): Tenant {
  return { id: r.id, name: r.name, apiKey: r.apiKey, createdAt: toIso(r.createdAt) };
}
function toUser(r: UserRow): User {
  return { id: r.id, email: r.email, name: r.name ?? "", apiKey: r.apiKey, createdAt: toIso(r.createdAt) };
}
function toMembership(r: MembershipRow): Membership {
  return { tenantId: r.tenantId, userId: r.userId, role: r.role as MemberRole };
}

export class TenantRepository {
  async get(tenantId: string): Promise<Tenant | null> {
    const rows = await db.select().from(tenants).where(eq(tenants.id, tenantId)).limit(1);
    return rows[0] ? toTenant(rows[0]) : null;
  }

  async getByApiKey(apiKey: string): Promise<Tenant | null> {
    const rows = await db.select().from(tenants).where(eq(tenants.apiKey, apiKey)).limit(1);
    return rows[0] ? toTenant(rows[0]) : null;
  }

  async list(): Promise<Tenant[]> {
    const rows = await db.select().from(tenants).orderBy(asc(tenants.createdAt));
    return rows.map(toTenant);
  }

  async exists(tenantId: string): Promise<boolean> {
    const rows = await db.select({ id: tenants.id }).from(tenants).where(eq(tenants.id, tenantId)).limit(1);
    return rows.length > 0;
  }

  async insert(tenantId: string, name: string, apiKey: string): Promise<void> {
    await db.insert(tenants).values({ id: tenantId, name, apiKey });
  }

  async delete(tenantId: string): Promise<boolean> {
    const deleted = await db.delete(tenants).where(eq(tenants.id, tenantId)).returning({ id: tenants.id });
    return deleted.length > 0;
  }
}

export class UserRepository {
  async get(userId: string): Promise<User | null> {
    const rows = await db.select().from(users).where(eq(users.id, userId)).limit(1);
    return rows[0] ? toUser(rows[0]) : null;
  }

  async getByApiKey(apiKey: string): Promise<User | null> {
    const rows = await db.select().from(users).where(eq(users.apiKey, apiKey)).limit(1);
    return rows[0] ? toUser(rows[0]) : null;
  }

  async getByEmail(email: string): Promise<User | null> {
    const rows = await db.select().from(users).where(eq(users.email, email)).limit(1);
    return rows[0] ? toUser(rows[0]) : null;
  }

  async insert(userId: string, email: string, name: string, apiKey: string): Promise<void> {
    await db.insert(users).values({ id: userId, email, name, apiKey });
  }
}

export class MembershipRepository {
  async get(tenantId: string, userId: string): Promise<Membership | null> {
    const rows = await db
      .select()
      .from(memberships)
      .where(and(eq(memberships.tenantId, tenantId), eq(memberships.userId, userId)))
      .limit(1);
    return rows[0] ? toMembership(rows[0]) : null;
  }

  async listForTenant(tenantId: string): Promise<Membership[]> {
    const rows = await db
      .select()
      .from(memberships)
      .where(eq(memberships.tenantId, tenantId))
      .orderBy(asc(memberships.role), asc(memberships.userId));
    return rows.map(toMembership);
  }

  async upsert(tenantId: string, userId: string, role: string): Promise<void> {
    await db
      .insert(memberships)
      .values({ tenantId, userId, role })
      .onConflictDoUpdate({ target: [memberships.tenantId, memberships.userId], set: { role } });
  }

  async delete(tenantId: string, userId: string): Promise<boolean> {
    const deleted = await db
      .delete(memberships)
      .where(and(eq(memberships.tenantId, tenantId), eq(memberships.userId, userId)))
      .returning({ userId: memberships.userId });
    return deleted.length > 0;
  }

  async countOwners(tenantId: string): Promise<number> {
    const rows = await db
      .select({ c: count() })
      .from(memberships)
      .where(and(eq(memberships.tenantId, tenantId), eq(memberships.role, "owner")));
    return Number(rows[0]?.c ?? 0);
  }
}
